src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
> Also, the locks need to go in (be shared between) various places, 
> including the implementation of OWLOntologyManager. It's my (current) 
> understanding that the lock for a manager should be the same for its 
> ontologies as well (well, makes sense - @stdotjohn 
> <https://github.com/stdotjohn> is this correct?).
>

Yes that is correct.  At one time I had lots of different reader-writer 
locks but that was far too complicated.  So I changed it to the less 
fine grained idea of using one reader-writer lock associated with the 
manager.

> Then I think extra locks are needed for lists of listeners 
> (adding/removing a listener requires a read lock to be acquired I think).
>

Something like this sounds right.  I don't see that code in the current 
version and I am not sure why.  It has been a very long time and also I 
am not up on what has changed.

-Timothy


On 04/08/2015 02:16 PM, Matthew Horridge wrote:
>
> In 
> src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java 
> <https://github.com/protegeproject/protege-owlapi/pull/3#discussion_r28014289>:
>
> > +        }
> > +    }
> > +
> > +    @Override
> > +    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
> > +        writeLock.lock();
> > +        try {
> > +            delegate.setOWLOntologyManager(saveOntology);
> > +        } finally {
> > +            writeLock.unlock();
> > +        }
> > +    }
> > +
> > +    @Override
> > +    public void saveOntology() throws OWLOntologyStorageException {
> > +        writeLock.lock();
>
>     Speaking of which, I think we can reach the same effect putting
>     read write locks around Internals. Smaller API so easier to maintain.
>
> Perhaps. Mucking about with the internals in the future might not be 
> so easy though. The one thing with the current code (ignoring the 
> design of how the factories etc. fit together, but this is an OWL API 
> thing anyway) is that it's pretty simple to see what is going on with 
> locking in each ontology. i.e. it's immediately possible to see that 
> gettingX requires a read lock be acquired. I think this is a plus.
>
> Also, the locks need to go in (be shared between) various places, 
> including the implementation of OWLOntologyManager. It's my (current) 
> understanding that the lock for a manager should be the same for its 
> ontologies as well (well, makes sense - @stdotjohn 
> <https://github.com/stdotjohn> is this correct?). Then I think extra 
> locks are needed for lists of listeners (adding/removing a listener 
> requires a read lock to be acquired I think).
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/protegeproject/protege-owlapi/pull/3/files#r28014289>.
>